Warp IO failed due to "We encountered an internal error"

### Environment info
- NooBaa Version: master-20210926
- Platform: OCP 4.8.9

### Actual behavior
1.Warp IO failed due to error - We encountered an internal error

### Expected behavior
1. IO shouldn't fail

### Steps to reproduce
1. Started IO using warp command: 
`warp mixed --access-key=ajUiz6hYeDFGZizM4Rk0 --secret-key=0K1loazPi+GZEtvDdK8e/N1BMxIAacjaiU7GNWI7 --obj.size=1G --host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com --duration=1h --bucket=bucket-3 --debug --tls --insecure`

Got multiple errors and IO failed. 

Logs from endpoint pod:

```
Sep-27 12:38:53.438 [Endpoint/14] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-3/JNbabmS3/1.oUT%29xacUUj41yMOS.rnd?partNumber=1&amp;uploadId=1fcd9282-5315-4687-ab69-113036f77004</Resource><RequestId>ku2mxt55-dh5x7x-33i</RequestId></Error> PUT /bucket-3/JNbabmS3/1.oUT%29xacUUj41yMOS.rnd?partNumber=1&uploadId=1fcd9282-5315-4687-ab69-113036f77004 {"user-agent":"MinIO (linux; amd64) minio-go/v7.0.11 warp/0.3.43","content-length":"134217728","authorization":"AWS4-HMAC-SHA256 Credential=ajUiz6hYeDFGZizM4Rk0/20210927/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=5d4af550e4a8d8294e1bdd6905f7cb4abbf2e2d9ad6913c1aeddde4f36081ba8","x-amz-content-sha256":"UNSIGNED-PAYLOAD","x-amz-date":"20210927T123653Z","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: Semaphore Worker Timeout
    at /root/node_modules/noobaa-core/src/util/semaphore.js:40:25
    at new Promise (<anonymous>)
    at Semaphore._work_cap (/root/node_modules/noobaa-core/src/util/semaphore.js:39:16)
    at Semaphore.surround_count (/root/node_modules/noobaa-core/src/util/semaphore.js:77:47)
    at async NamespaceFS.upload_multipart (/root/node_modules/noobaa-core/src/sdk/namespace_fs.js:815:30)
    at async Object.put_object_uploadId [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/ops/s3_put_object_uploadId.js:31:17)
    at async handle_request (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:149:19)
    at async Object.s3_rest [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:68:9)
Sep-27 12:38:53.444 [Endpoint/14] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-3/3Bu1ilRl/1.iJs78turdr13HjRv.rnd?partNumber=1&amp;uploadId=dccae3a3-5c7b-4d43-afa8-1d5bc908bc9e</Resource><RequestId>ku2mxt58-dim4n2-s25</RequestId></Error> PUT /bucket-3/3Bu1ilRl/1.iJs78turdr13HjRv.rnd?partNumber=1&uploadId=dccae3a3-5c7b-4d43-afa8-1d5bc908bc9e {"user-agent":"MinIO (linux; amd64) minio-go/v7.0.11 warp/0.3.43","content-length":"134217728","authorization":"AWS4-HMAC-SHA256 Credential=ajUiz6hYeDFGZizM4Rk0/20210927/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=ddc2ad219d8e7bd29fd5ee7cde268a2dc44d30d82987317d59ce0598738fa1b1","x-amz-content-sha256":"UNSIGNED-PAYLOAD","x-amz-date":"20210927T123653Z","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: Semaphore Worker Timeout
    at /root/node_modules/noobaa-core/src/util/semaphore.js:40:25
    at new Promise (<anonymous>)
    at Semaphore._work_cap (/root/node_modules/noobaa-core/src/util/semaphore.js:39:16)
    at Semaphore.surround_count (/root/node_modules/noobaa-core/src/util/semaphore.js:77:47)
    at async NamespaceFS.upload_multipart (/root/node_modules/noobaa-core/src/sdk/namespace_fs.js:815:30)
    at async Object.put_object_uploadId [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/ops/s3_put_object_uploadId.js:31:17)
    at async handle_request (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:149:19)
    at async Object.s3_rest [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:68:9)
```

Other warp commands in parallel are working well. They are uploading objects of 9M, 100M. 

### More information - Screenshots / Logs / Other output

Hi @jeniawhite , Similar problem was seen this morning while doing file upload of 100GB. There was in parallel file upload of 10GB to the same bucket too which didn't fail. 

`upload failed: ./big_file_100GB to s3://bucket-51/big_file_100GB An error occurred () when calling the UploadPart operation:`

```
Sep-30 4:33:55.419 [Endpoint/13] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Pleas
e try again.</Message><Resource>/bucket-51/big_file_100GB?uploadId=75cbea55-565a-4112-845e-4afd436308d7&amp;partNumber=336</Resource><RequestId>ku6fxozp-6w426e-1016</RequestId></Error> PUT /bucket
-51/big_file_100GB?uploadId=75cbea55-565a-4112-845e-4afd436308d7&partNumber=336 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.19.1.el8_4.x86_64 exe/x86_6
4.rhel.8 prompt/off command/s3.cp","content-md5":"PaKchm04tjgNqENrnBZCQw==","expect":"100-continue","x-amz-date":"20210930T043154Z","x-amz-content-sha256":"UNSIGNED-PAYLOAD","authorization":"AWS4-
HMAC-SHA256 Credential=BrVzxXwyhi470cqA3s0c/20210930/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=d7dceb1114db3b1243dc3841a0e833578dfa1e1fdf
9990c55f5982dde2c429d5","content-length":"16777216","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","
x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: Semaphore Worker Timeout
    at /root/node_modules/noobaa-core/src/util/semaphore.js:40:25
    at new Promise (<anonymous>)
    at Semaphore._work_cap (/root/node_modules/noobaa-core/src/util/semaphore.js:39:16)
    at Semaphore.surround_count (/root/node_modules/noobaa-core/src/util/semaphore.js:77:47)
    at async NamespaceFS.upload_multipart (/root/node_modules/noobaa-core/src/sdk/namespace_fs.js:815:30)
    at async Object.put_object_uploadId [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/ops/s3_put_object_uploadId.js:31:17)
    at async handle_request (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:149:19)
    at async Object.s3_rest [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:68:9)
Sep-30 4:33:55.419 [Endpoint/13] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Pleas
e try again.</Message><Resource>/bucket-51/big_file_100GB?uploadId=75cbea55-565a-4112-845e-4afd436308d7&amp;partNumber=294</Resource><RequestId>ku6fxozs-6xukdt-rly</RequestId></Error> PUT /bucket-
51/big_file_100GB?uploadId=75cbea55-565a-4112-845e-4afd436308d7&partNumber=294 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.19.1.el8_4.x86_64 exe/x86_64
.rhel.8 prompt/off command/s3.cp","content-md5":"rL9/SnjTcmysAcezLFZiSA==","expect":"100-continue","x-amz-date":"20210930T043154Z","x-amz-content-sha256":"UNSIGNED-PAYLOAD","authorization":"AWS4-H
MAC-SHA256 Credential=BrVzxXwyhi470cqA3s0c/20210930/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=dd5600eec399a0f228c699bcfdcd655654ea98e0ec8
28491ee91060c86a698fd","content-length":"16777216","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x
-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: Semaphore Worker Timeout
    at /root/node_modules/noobaa-core/src/util/semaphore.js:40:25
    at new Promise (<anonymous>)
    at Semaphore._work_cap (/root/node_modules/noobaa-core/src/util/semaphore.js:39:16)
    at Semaphore.surround_count (/root/node_modules/noobaa-core/src/util/semaphore.js:77:47)
    at async NamespaceFS.upload_multipart (/root/node_modules/noobaa-core/src/sdk/namespace_fs.js:815:30)
    at async Object.put_object_uploadId [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/ops/s3_put_object_uploadId.js:31:17)
    at async handle_request (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:149:19)
    at async Object.s3_rest [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:68:9)
Sep-30 4:33:55.420 [Endpoint/13] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Pleas
e try again.</Message><Resource>/bucket-51/big_file_100GB?uploadId=75cbea55-565a-4112-845e-4afd436308d7&amp;partNumber=332</Resource><RequestId>ku6fxozr-6x2wl6-zjp</RequestId></Error> PUT /bucket-
51/big_file_100GB?uploadId=75cbea55-565a-4112-845e-4afd436308d7&partNumber=332 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.19.1.el8_4.x86_64 exe/x86_64
.rhel.8 prompt/off command/s3.cp","content-md5":"0dnws1NyL4w+SGMffpkunA==","expect":"100-continue","x-amz-date":"20210930T043154Z","x-amz-content-sha256":"UNSIGNED-PAYLOAD","authorization":"AWS4-H
MAC-SHA256 Credential=BrVzxXwyhi470cqA3s0c/20210930/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=5e9a6017fae7106e3c284a5a610123d356265bde4ae
291807dbdfbcfa5a4788f","content-length":"16777216","host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x
-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: Semaphore Worker Timeout
    at /root/node_modules/noobaa-core/src/util/semaphore.js:40:25
    at new Promise (<anonymous>)
    at Semaphore._work_cap (/root/node_modules/noobaa-core/src/util/semaphore.js:39:16)
    at Semaphore.surround_count (/root/node_modules/noobaa-core/src/util/semaphore.js:77:47)
    at async NamespaceFS.upload_multipart (/root/node_modules/noobaa-core/src/sdk/namespace_fs.js:815:30)
    at async Object.put_object_uploadId [as handler] (/root/node_modules/noobaa-core/src/endpoint/s3/ops/s3_put_object_uploadId.js:31:17)
    at async handle_request (/root/node_modules/noobaa-core/src/endpoint/s3/s3_rest.js:149:19)

```

--- a/src/sdk/namespace_fs.js
+++ b/src/sdk/namespace_fs.js
@@ -724,31 +724,39 @@ class NamespaceFS {
             target_file = await nb_native().fs.open(fs_account_config, upload_path, 'w', get_umasked_mode(config.BASE_MODE_FILE));
             const flush = async () => {
                 if (q_buffers.length) {
+                    if (MD5Async) {
+                        for await (const buf of q_buffers) await MD5Async.update(buf);
+                    }
                     await target_file.writev(fs_account_config, q_buffers);
                     q_buffers = [];
                     q_size = 0;
                 }
             };
-            let count = 1;
-            for await (let data of source_stream) {
-                if (MD5Async) await MD5Async.update(data);
-                // Update the write stats
-                stats_collector.instance(rpc_client).update_namespace_write_stats({
-                    namespace_resource_id: this.namespace_resource_id,
-                    size: data.length,
-                    count
+            // Not using async iterators with ReadableStreams due to unsettled promises issues on abort/destroy
+            await new Promise((resolve, reject) => {
+                let count = 1;
+                source_stream.on('data', async data => {
+                    // Update the write stats
+                    stats_collector.instance(rpc_client).update_namespace_write_stats({
+                        namespace_resource_id: this.namespace_resource_id,
+                        size: data.length,
+                        count
+                    });
+                    // clear count for next updates
+                    count = 0;
+                    while (data && data.length) {
+                        const available_size = config.NSFS_BUF_SIZE - q_size;
+                        const buf = (available_size < data.length) ? data.slice(0, available_size) : data;
+                        q_buffers.push(buf);
+                        q_size += buf.length;
+                        if (q_size === config.NSFS_BUF_SIZE) await flush();
+                        data = (available_size < data.length) ? data.slice(available_size) : null;
+                    }
                 });
-                // clear count for next updates
-                count = 0;
-                while (data && data.length) {
-                    const available_size = config.NSFS_BUF_SIZE - q_size;
-                    const buf = (available_size < data.length) ? data.slice(0, available_size) : data;
-                    q_buffers.push(buf);
-                    q_size += buf.length;
-                    if (q_size === config.NSFS_BUF_SIZE) await flush();
-                    data = (available_size < data.length) ? data.slice(available_size) : null;
-                }
-            }
+                source_stream.on('end', resolve);
+                source_stream.on('close', () => reject(new Error('_upload_stream: source_stream closed')));
+                source_stream.on('error', err => reject(err));
+            });
             await flush();
             if (MD5Async) {
                 fs_xattr = this._assign_md5_to_fs_xattr((await MD5Async.digest()).toString('hex'), fs_xattr);

--- a/src/test/unit_tests/test_namespace_fs.js
+++ b/src/test/unit_tests/test_namespace_fs.js
@@ -66,7 +66,6 @@ mocha.describe('namespace_fs', function() {
         });
 
         mocha.it('list src dir without delimiter', async function() {
-            nb_native().fs.set_debug_level(5);
             const res = await ns_src.list_objects({
                 bucket: src_bkt,
             }, dummy_object_sdk);

--- a/src/test/unit_tests/test_semaphore.js
+++ b/src/test/unit_tests/test_semaphore.js
@@ -143,7 +143,7 @@ mocha.describe('semaphore', function() {
         assert.strictEqual(sem.value, 1);
         try {
             await sem.surround_count(1, async function() {
-                return new Promise((resolve, reject) => setTimeout(resolve, 2));
+                return new Promise((resolve, reject) => setTimeout(resolve, 500));
             });
             throw new Error('Semaphore did not throw an error on non settled worker');
         } catch (error) {

--- a/src/util/semaphore.js
+++ b/src/util/semaphore.js
@@ -4,8 +4,6 @@
 const WaitQueue = require('./wait_queue');
 const dbg = require('./debug_module')(__filename);
 
-const DEFAULT_WORK_TIMEOUT = 2 * 60 * 1000;
-
 class Semaphore {
 
     /**
@@ -25,13 +23,12 @@ class Semaphore {
             this._verbose = Boolean(params && params.verbose);
             if (params.timeout) {
                 this._timeout = params.timeout;
+                this._timeout_error_code = params.timeout_error_code || 'SEMAPHORE_TIMEOUT';
             }
             if (params.work_timeout) {
                 this._work_timeout = params.work_timeout;
+                this._work_timeout_error_code = params.work_timeout_error_code || 'SEMAPHORE_WORKER_TIMEOUT';
             }
-            this._timeout_error_code = params.timeout_error_code || 'SEMAPHORE_TIMEOUT';
-            this._work_timeout_error_code = params.work_timeout_error_code || 'SEMAPHORE_WORKER_TIMEOUT';
-
         }
     }
 
@@ -39,7 +36,7 @@ class Semaphore {
         return new Promise((resolve, reject) => {
             const err = new Error('Semaphore Worker Timeout');
             err.code = this._work_timeout_error_code;
-            const t = setTimeout(() => reject(err), this._work_timeout || DEFAULT_WORK_TIMEOUT);
+            const t = setTimeout(() => reject(err), this._work_timeout);
             t.unref();
         });
     }
@@ -54,7 +51,7 @@ class Semaphore {
         await this.wait();
         try {
             // May lead to unaccounted work in the background on timeout
-            return Promise.race([func(), this._work_cap()]);
+            return this._work_timeout ? Promise.race([func(), this._work_cap()]) : await func();
         } finally {
             // Release should be called only when the wait was successful
             // If the item did not take any "resources"/value from the semaphore
@@ -74,7 +71,7 @@ class Semaphore {
         await this.wait(count);
         try {
             // May lead to unaccounted work in the background on timeout
-            return Promise.race([func(), this._work_cap()]);
+            return this._work_timeout ? Promise.race([func(), this._work_cap()]) : await func();
         } finally {
             // Release should be called only when the wait was successful
             // If the item did not take any "resources"/value from the semaphore
@@ -166,6 +163,7 @@ class Semaphore {
                 () => this._on_timeout(),
                 this._timeout
             );
+            this._timer.unref();
         }
 
         this._waiting_value += count;

